src/api/documents/document-ref.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  DocumentServiceClient,
+  DocumentGetRequest,
+  DocumentGetResponse,
+  Key,
+  Collection as DocCollection,
+  DocumentSetRequest,
+  DocumentDeleteRequest,
+} from '../../interfaces/document';
+import { CollectionRef } from './collection-ref';
+import { getKey } from './key';
+
+/**
+ * Document Ref
+ *
+ * Provides a Document Reference class.
+ * Used to create references to collections.
+ */
+export class DocumentRef<T extends { [key: string]: any }> {
+  private documentClient: DocumentServiceClient;
+  private key: Key;
+
+  constructor(
+    documentClient: DocumentServiceClient,
+    collection: DocCollection,
+    documentId: string
+  ) {
+    this.documentClient = documentClient;
+    this.key = getKey(collection, documentId);
+  }
+
+  /**
+   * Return the collection document reference value.
+   * @returns the collection document reference value, or null if not found
+   */
+  public async get() {
+    const request = new DocumentGetRequest();
+    request.setKey(this.key);
+
+    return new Promise<T>((resolve, reject) => {
+      this.documentClient.get(
+        request,
+        (error, response: DocumentGetResponse) => {
+          if (error) {
+            reject(error);
+          } else if (response.hasDocument()) {
+            const document = response.getDocument();
+            const content = document.getContent().toJavaScript() as T;
+
+            resolve(content);
+          } else {
+            resolve(null);
+          }
+        }
+      );
+    });
+  }
+
+  /**
+   * Set the document content for this document reference in the database. If the
+   * document does not exist an new item will be created, otherwise an
+   * existing document will be update with the new value.
+   * @param value content the document content to store (required)
+   */
+  public async set(value: T) {
+    const request = new DocumentSetRequest();
+    request.setKey(this.key);
+    request.setContent(Struct.fromJavaScript(value));
+
+    return new Promise<void>((resolve, reject) => {
+      this.documentClient.set(request, (error) => {
+        if (error) {
+          reject(error);
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
+
+  /**
+   * Delete this document reference from the database if it exists.
+   */
+  public async delete() {
+    const request = new DocumentDeleteRequest();
+    request.setKey(this.key);
+
+    return new Promise<void>((resolve, reject) => {
+      this.documentClient.delete(request, (error) => {
+        if (error) {
+          reject(error);
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
+
+  /**
+   * Gets a Collection instance that refers to the collection at the specified path.
+   * @param name The name of the collection (required)
+   * @returns The Collection instance
+   */
+  public collection<T extends { [key: string]: any }>(name: string) {
+    return new CollectionRef<T>(this.documentClient, name, this.key);
```

Comment:
should we be throwing an error here if our current max depth of 1 collection is exceeded.